hail/python/hailtop/batch_client/aioclient.py

```diff
@@ -305,6 +305,217 @@ async def attempts(self):
         return await resp.json()
 
 
+class AbsoluteJobGroupId(int):
+    pass
+
+
+class InUpdateJobGroupId(int):
+    pass
+
+
+class JobGroupAlreadySubmittedError(Exception):
+    pass
+
+
+class JobGroupNotSubmittedError(Exception):
+    pass
+
+
+class JobGroup:
+    @staticmethod
+    def submitted_job_group(
+        batch: 'Batch',
+        job_group_id: int,
+        *,
+        _attributes: Optional[Dict[str, str]] = None,
+        _last_known_status: Optional[dict] = None,
+    ) -> 'JobGroup':
+        return JobGroup(
+            batch, AbsoluteJobGroupId(job_group_id), _attributes=_attributes, _last_known_status=_last_known_status
+        )
+
+    @staticmethod
+    def unsubmitted_job_group(batch: 'Batch', job_group_id: int, *, attributes: Optional[Dict[str, str]]) -> 'JobGroup':
+        return JobGroup(batch, InUpdateJobGroupId(job_group_id), _attributes=attributes)
+
+    def __init__(
+        self,
+        batch: 'Batch',
+        job_group_id: Union[AbsoluteJobGroupId, InUpdateJobGroupId],
+        *,
+        _attributes: Optional[Dict[str, str]] = None,
+        _last_known_status: Optional[dict] = None,
+    ):
+        self._batch = batch
+        self._job_group_id = job_group_id
+
+        self._attributes = _attributes or {}
+        self._last_known_status = _last_known_status
+
+    def _submit(self, in_update_start_job_group_id: int):
+        self._raise_if_submitted()
+        self._job_group_id = AbsoluteJobGroupId(in_update_start_job_group_id + self._job_group_id - 1)
```

Comment:
Erm. I see we're already doing this with job ids. Hmm. OK. Ignore.

Reply:
Basically every change mirrored either the existing `Job` code or the `Batch` code. If there's an issue with the design, I'm happy to change it, but it will be changed in both places.